### ISSUE TITLE
Win 7: assertion failed: validate_hidpi_factor(dpi_factor)

### ISSUE THREAD

**Issue by sinesc** [tomaka/winit#607]
I am getting frequent validate_hidpi_factor assertion failure panics on Windows 7 in dpi.rs in line 273:9 and 153:9 (random which) using winit 0.16.2. This does not seem to occur on Windows 10 or Linux.

```
thread 'main' panicked at 'assertion failed: validate_hidpi_factor(dpi_factor)', C:\Users\sinesc\.cargo\registry\src\github.com-1ecc6299db9ec823\winit-0.16.2\src\dpi.rs:273:9
stack backtrace:
   0:        0x14018cad3 - std::sys::windows::backtrace::unwind_backtrace
                               at C:\projects\rust\src\libstd\sys\windows\backtrace\mod.rs:95
   1:        0x14019676e - std::sys_common::backtrace::print
                               at C:\projects\rust\src\libstd\sys_common\backtrace.rs:59
   2:        0x140186758 - std::panicking::default_hook::{{closure}}
                               at C:\projects\rust\src\libstd\panicking.rs:211
   3:        0x140186411 - std::panicking::default_hook
                               at C:\projects\rust\src\libstd\panicking.rs:227
   4:        0x140186f5c - std::panicking::rust_panic_with_hook
                               at C:\projects\rust\src\libstd\panicking.rs:475
   5:        0x13fd656c2 - std::panicking::begin_panic<str*>
                               at C:\projects\rust\src\libstd\panicking.rs:409
   6:        0x13fd362a3 - winit::dpi::PhysicalSize::to_logical
                               at C:\Users\sinesc\.cargo\registry\src\github.com-1ecc6299db9ec823\winit-0.16.2\src\dpi.rs:273
   7:        0x13fd35fca - winit::dpi::LogicalSize::from_physical<(u32, u32)>
                               at C:\Users\sinesc\.cargo\registry\src\github.com-1ecc6299db9ec823\winit-0.16.2\src\dpi.rs:208
   8:        0x13fbff8c7 - winit::platform::platform::window::{{impl}}::get_inner_size::{{closure}}
                               at C:\Users\sinesc\.cargo\registry\src\github.com-1ecc6299db9ec823\winit-0.16.2\src\platform\windows\window.rs:194
   9:        0x13fbf1570 - core::option::Option<(u32, u32)>::map<(u32, u32),winit::dpi::LogicalSize,closure>
                               at C:\projects\rust\src\libcore\option.rs:414
  10:        0x13fbff885 - winit::platform::platform::window::Window::get_inner_size
                               at C:\Users\sinesc\.cargo\registry\src\github.com-1ecc6299db9ec823\winit-0.16.2\src\platform\windows\window.rs:191
  11:        0x13fca19aa - winit::Window::get_inner_size
                               at C:\Users\sinesc\.cargo\registry\src\github.com-1ecc6299db9ec823\winit-0.16.2\src\window.rs:252
  12:        0x13fc54448 - glium::backend::glutin::{{impl}}::get_framebuffer_dimensions
                               at C:\Users\sinesc\.cargo\registry\src\github.com-1ecc6299db9ec823\glium-0.22.0\src\backend\glutin\mod.rs:262
  13:        0x13fb8cabd - glium::context::Context::get_framebuffer_dimensions
                               at C:\Users\sinesc\.cargo\registry\src\github.com-1ecc6299db9ec823\glium-0.22.0\src\context\mod.rs:238
  14:        0x13fb8c73a - glium::backend::glutin::Display::draw
                               at C:\Users\sinesc\.cargo\registry\src\github.com-1ecc6299db9ec823\glium-0.22.0\src\backend\glutin\mod.rs:171
  15:        0x13faad329 - radiant_rs::backends::glium::Display::draw
                               at C:\Users\sinesc\Projekte\radiant-rs\src\backends\glium.rs:192
  16:        0x13fa399a8 - radiant_rs::core::display::Display::prepare_frame
                               at C:\Users\sinesc\Projekte\radiant-rs\src\core\display.rs:79
  17:        0x13fa39b78 - radiant_rs::core::display::Display::clear_frame
                               at C:\Users\sinesc\Projekte\radiant-rs\src\core\display.rs:84
  18:        0x13fa21a6f - demo_blobs::main::{{closure}}
                               at C:\Users\sinesc\Projekte\radiant-rs\examples\demo_blobs.rs:43
  19:        0x13fa25f2e - radiant_utils::misc::loops::renderloop<closure>
                               at C:\Users\sinesc\.cargo\registry\src\github.com-1ecc6299db9ec823\radiant-utils-0.3.0\src\misc\loops.rs:114
  20:        0x13fa23f9a - demo_blobs::main
                               at C:\Users\sinesc\Projekte\radiant-rs\examples\demo_blobs.rs:42
  21:        0x13fa2553f - std::rt::lang_start::{{closure}}<()>
                               at C:\projects\rust\src\libstd\rt.rs:74
  22:        0x140186926 - std::panicking::try::do_call<closure,i32>
                               at C:\projects\rust\src\libstd\panicking.rs:310
  23:        0x14019a171 - panic_unwind::__rust_maybe_catch_panic
                               at C:\projects\rust\src\libpanic_unwind\lib.rs:106
  24:        0x140184c82 - std::rt::lang_start_internal
                               at C:\projects\rust\src\libstd\rt.rs:58
  25:        0x13fa2551a - std::rt::lang_start<()>
                               at C:\projects\rust\src\libstd\rt.rs:74
  26:        0x13fa2434f - main
  27:        0x1401a9b63 - __scrt_common_main_seh
                               at f:\dd\vctools\crt\vcstartup\src\startup\exe_common.inl:283
  28:         0x777159cc - BaseThreadInitThunk
  29:         0x7787383c - RtlUserThreadStart
```

```
thread '<unnamed>' panicked at 'assertion failed: validate_hidpi_factor(dpi_factor)', C:\Users\sinesc\.cargo\registry\src\github.com-1ecc6299db9ec823\winit-0.16.2\src\dpi.rs:153:9
stack backtrace:
   0:        0x13f97cad3 - std::sys::windows::backtrace::unwind_backtrace
                               at C:\projects\rust\src\libstd\sys\windows\backtrace\mod.rs:95
   1:        0x13f98676e - std::sys_common::backtrace::print
                               at C:\projects\rust\src\libstd\sys_common\backtrace.rs:59
   2:        0x13f976758 - std::panicking::default_hook::{{closure}}
                               at C:\projects\rust\src\libstd\panicking.rs:211
   3:        0x13f976411 - std::panicking::default_hook
                               at C:\projects\rust\src\libstd\panicking.rs:227
   4:        0x13f976f5c - std::panicking::rust_panic_with_hook
                               at C:\projects\rust\src\libstd\panicking.rs:475
   5:        0x13f5556c2 - std::panicking::begin_panic<str*>
                               at C:\projects\rust\src\libstd\panicking.rs:409
   6:        0x13f525dd3 - winit::dpi::PhysicalPosition::to_logical
                               at C:\Users\sinesc\.cargo\registry\src\github.com-1ecc6299db9ec823\winit-0.16.2\src\dpi.rs:153
   7:        0x13f525cd2 - winit::dpi::LogicalPosition::from_physical<(f64, f64)>
                               at C:\Users\sinesc\.cargo\registry\src\github.com-1ecc6299db9ec823\winit-0.16.2\src\dpi.rs:88
   8:        0x13f536f42 - winit::platform::platform::events_loop::callback
                               at C:\Users\sinesc\.cargo\registry\src\github.com-1ecc6299db9ec823\winit-0.16.2\src\platform\windows\events_loop.rs:593
   9:         0x77619bbc - TranslateMessageEx
  10:         0x77613b93 - CallWindowProcW
  11:         0x77613b0f - CallWindowProcW
  12:      0x7feda12948d - glDebugEntry
  13:         0x77619bbc - TranslateMessageEx
  14:         0x776198c1 - TranslateMessage
  15:        0x13f534c17 - winit::platform::platform::events_loop::{{impl}}::with_dpi_awareness::{{closure}}
                               at C:\Users\sinesc\.cargo\registry\src\github.com-1ecc6299db9ec823\winit-0.16.2\src\platform\windows\events_loop.rs:197
  16:        0x13f548d50 - std::sys_common::backtrace::__rust_begin_short_backtrace<closure,()>
                               at C:\projects\rust\src\libstd\sys_common\backtrace.rs:136
  17:        0x13f555640 - std::thread::{{impl}}::spawn::{{closure}}::{{closure}}<closure,()>
                               at C:\projects\rust\src\libstd\thread\mod.rs:409
  18:        0x13f554b60 - std::panic::{{impl}}::call_once<(),closure>
                               at C:\projects\rust\src\libstd\panic.rs:313
  19:        0x13f5558ad - std::panicking::try::do_call<std::panic::AssertUnwindSafe<closure>,()>
                               at C:\projects\rust\src\libstd\panicking.rs:310
  20:        0x13f98a171 - panic_unwind::__rust_maybe_catch_panic
                               at C:\projects\rust\src\libpanic_unwind\lib.rs:106
  21:        0x13f555789 - std::panicking::try<(),std::panic::AssertUnwindSafe<closure>>
                               at C:\projects\rust\src\libstd\panicking.rs:289
  22:        0x13f554d60 - std::panic::catch_unwind<std::panic::AssertUnwindSafe<closure>,()>
                               at C:\projects\rust\src\libstd\panic.rs:392
  23:        0x13f555493 - std::thread::{{impl}}::spawn::{{closure}}<closure,()>
                               at C:\projects\rust\src\libstd\thread\mod.rs:408
  24:        0x13f558e69 - alloc::boxed::{{impl}}::call_box<(),closure>
                               at C:\projects\rust\src\liballoc\boxed.rs:640
  25:        0x13f977521 - std::sys::windows::thread::{{impl}}::new::thread_start
                               at C:\projects\rust\src\libstd\sys\windows\thread.rs:55
  26:         0x777159cc - BaseThreadInitThunk
  27:         0x7787383c - RtlUserThreadStart
````

Is there any other information that might be helpful?

**Comment by francesca64** [comment]
If you add a line to `validate_hidpi_factor` that prints the DPI factor, what's the value?

**Comment by sinesc** [comment]
I modified the visa/7 case in get_window_dpi like this:

```rust
// We are on Vista or later.
if winuser::IsProcessDPIAware() != FALSE {
    // If the process is DPI aware, then scaling must be handled by the application using
    // this DPI value.
    use std::time::Instant;
    use winapi::um::wingdi::LOGPIXELSY;
    let now = Instant::now();
    let sx = GetDeviceCaps(hdc, LOGPIXELSX);
    println!("{:?} getDeviceCaps({:?}, LOGPIXELSX)={:?}", now, hdc, sx);
    if sx == 0 {
        let sy = GetDeviceCaps(hdc, LOGPIXELSY);
        println!("{:?} getDeviceCaps({:?}, LOGPIXELSY)={:?}", now, hdc, sy);
    }
    sx as u32
} else {
    // If the process is DPI unaware, then scaling is performed by the OS; we thus return
    // 96 (scale factor 1.0) to prevent the window from being re-scaled by both the
    // application and the WM.
    BASE_DPI
}
```

This results in the following output:

```
[..]
Instant { t: 4850410442 } getDeviceCaps(0x26010f5e, LOGPIXELSX)=0
Instant { t: 4850410442 } getDeviceCaps(0x26010f5e, LOGPIXELSY)=96
[..]
```

Including trace:

```
[...]
Instant { t: 4850344483 } getDeviceCaps(0x26010f5e, LOGPIXELSX)=96
Instant { t: 4850347500 } getDeviceCaps(0x26010f5e, LOGPIXELSX)=96
Instant { t: 4850408708 } getDeviceCaps(0x26010f5e, LOGPIXELSX)=96
Instant { t: 4850409919 } getDeviceCaps(0x26010f5e, LOGPIXELSX)=96
Instant { t: 4850410442 } getDeviceCaps(0x26010f5e, LOGPIXELSX)=0
Instant { t: 4850410442 } getDeviceCaps(0x26010f5e, LOGPIXELSY)=96
thread 'Instant { t: 4850410443 } getDeviceCaps(0x26010f5e, LOGPIXELSX)=96
main' panicked at 'assertion failed: validate_hidpi_factor(dpi_factor)', C:\Users\sinesc\.cargo\registry\src\github.com-1ecc6299db9ec823\winit-0.16.2\src\dpi.rs:213:9
Instant { t: 4850457070 } getDeviceCaps(0x26010f5e, LOGPIXELSX)=96
Instant { t: 4850472698 } getDeviceCaps(0x26010f5e, LOGPIXELSX)=96
Instant { t: 4850520036 } getDeviceCaps(0x26010f5e, LOGPIXELSX)=96
Instant { t: 4850535444 } getDeviceCaps(0x26010f5e, LOGPIXELSX)=96
Instant { t: 4850582228 } getDeviceCaps(0x26010f5e, LOGPIXELSX)=96
Instant { t: 4850597907 } getDeviceCaps(0x26010f5e, LOGPIXELSX)=96
stack backtrace:
Instant { t: 4850645036 } getDeviceCaps(0x26010f5e, LOGPIXELSX)=96
Instant { t: 4850660666 } getDeviceCaps(0x26010f5e, LOGPIXELSX)=96
   0:        0x14039cf13 - std::sys::windows::backtrace::unwind_backtrace
                               at C:\projects\rust\src\libstd\sys\windows\backtrace\mod.rs:95
   1:        0x1403a6c0e - std::sys_common::backtrace::print
                               at C:\projects\rust\src\libstd\sys_common\backtrace.rs:59
   2:        0x140396b98 - std::panicking::default_hook::{{closure}}
                               at C:\projects\rust\src\libstd\panicking.rs:211
   3:        0x140396851 - std::panicking::default_hook
                               at C:\projects\rust\src\libstd\panicking.rs:227
   4:        0x14039739c - std::panicking::rust_panic_with_hook
                               at C:\projects\rust\src\libstd\panicking.rs:475
   5:        0x13ff81ac2 - std::panicking::begin_panic<str*>
                               at C:\projects\rust\src\libstd\panicking.rs:409
   6:        0x13fe0f403 - winit::dpi::LogicalSize::to_physical
                               at \<panic macros>:3
   7:        0x13fe64212 - glium::backend::glutin::{{impl}}::get_framebuffer_dimensions::{{closure}}
                               at C:\Users\sinesc\.cargo\registry\src\github.com-1ecc6299db9ec823\glium-0.22.0\src\backend\glutin\mod.rs:263
   8:        0x13fdffcae - core::option::Option<winit::dpi::LogicalSize>::map<winit::dpi::LogicalSize,winit::dpi::PhysicalSize,closure>
                               at C:\projects\rust\src\libcore\option.rs:414
   9:     Instant { t: 4850707267 } getDeviceCaps(0x26010f5e, LOGPIXELSX)=96
   0x13fe64123 - glium::backend::glutin::{{impl}}::get_framebuffer_dimensions
                               at C:\Users\sinesc\.cargo\registry\src\github.com-1ecc6299db9ec823\glium-0.22.0\src\backend\glutin\mod.rs:262
  10:        0x13fd9c8ed - glium::context::Context::get_framebuffer_dimensions
                               at C:\Users\sinesc\.cargo\registry\src\github.com-1ecc6299db9ec823\glium-0.22.0\src\context\mod.rs:238
  11:        0x13fd9c56a - glium::backend::glutin::Display::draw
                               at C:\Users\sinesc\.cargo\registry\src\github.com-1ecc6299db9ec823\glium-0.22.0\src\backend\glutin\mod.rs:171
  12:        0x13fcbd329 - radiant_rs::backends::glium::Display::draw
                               at C:\Users\sinesc\Projekte\radiant-rs\src\backends\glium.rs:194
  13:     Instant { t: 4850722575 } getDeviceCaps(0x26010f5e, LOGPIXELSX)=96
   0x13fc499b8 - radiant_rs::core::display::Display::prepare_frame
                               at C:\Users\sinesc\Projekte\radiant-rs\src\core\display.rs:79
  14:        0x13fc49b88 - radiant_rs::core::display::Display::clear_frame
                               at C:\Users\sinesc\Projekte\radiant-rs\src\core\display.rs:84
  15:        0x13fc31a6f - demo_blobs::main::{{closure}}
                               at C:\Users\sinesc\Projekte\radiant-rs\examples\demo_blobs.rs:43
  16:        0x13fc35f3e - radiant_utils::misc::loops::renderloop<closure>
                               at C:\Users\sinesc\.cargo\registry\src\github.com-1ecc6299db9ec823\radiant-utils-0.3.0\src\misc\loops.rs:114
  17:        0x13fc33fb7 - demo_blobs::main
                               at C:\Users\sinesc\Projekte\radiant-rs\examples\demo_blobs.rs:42
  18:        0x13fc3554f - std::rt::lang_start::{{closure}}<()>
                               at C:\projects\rust\src\libstd\rt.rs:74
  19:        0x140396d66 - std::panicking::try::do_call<closure,i32>
                               at C:\projects\rust\src\libstd\panicking.rs:310
  20:        0x1403aa611 - panic_unwind::__rust_maybe_catch_panic
                               at C:\projects\rust\src\libpanic_unwind\lib.rs:106
  21:        0x140395062 - std::rt::lang_start_internal
                               at C:\projects\rust\src\libstd\rt.rs:58
  22:        0x13fc3552a - std::rt::lang_start<()>
                               at C:\projects\rust\src\libstd\rt.rs:74
  23:        0x13fc3435f - main
  24:        0x1403ba003 - __scrt_common_main_seh
                               at f:\dd\vctools\crt\vcstartup\src\startup\exe_common.inl:283
  25:         0x76d859cc - BaseThreadInitThunk
  26:         0x76fe383c - RtlUserThreadStart
```

**Comment by francesca64** [comment]
So, the solution is simply to call `GetDeviceCaps` on both `LOGPIXELSX` and `LOGPIXELSY`, and only use DPI values acquired after both have been called? Would you like to make a PR for this?

**Comment by sinesc** [comment]
I mentioned a little hidden below the stack trace that X/Y didn't actually matter, simply calling `GetDeviceCaps` a second time with `LOGPIXELSX` also "fixed" the issue. 

There is also this part in the trace that looks a little suspicious:

Instant { t: 4850409919 } getDeviceCaps(0x26010f5e, LOGPIXELSX)=96
Instant { t: 4850410442 } getDeviceCaps(0x26010f5e, LOGPIXELSX)=0
Instant { t: `4850410442` } getDeviceCaps(0x26010f5e, LOGPIXELSY)=96
**thread 'Instant { t: `4850410443` } getDeviceCaps(0x26010f5e, LOGPIXELSX)=96
main' panicked** at 'assertion failed: validate_hidpi_factor(dpi_factor)', C:\Users\sinesc\.cargo\registry\src\github.com-1ecc6299db9ec823\winit-0.16.2\src\dpi.rs:213:9
Instant { t: 4850457070 } getDeviceCaps(0x26010f5e, LOGPIXELSX)=96


The two calls seem to be coming from different threads (the test app was single-threaded) right on top of each other. The `Instant`s are only 1ns apart and one interrupts the trace output.

Not sure if simply calling `GetDeviceCaps` again would be an actual fix or just reduce the chance of this issue occuring, otherwise I'd make a PR.

**Comment by sinesc** [comment]
Ok, so I let the testcase run for 15 minutes without using the PC - it worked fine. Then I started moving the mouse inside the window - it panicked within seconds. I think it has to do with the event handling calling `getDeviceCaps` at the same time. The relevant `Instant`s are within one unit again. Does this make any sense?

Updated code:

```rust
// We are on Vista or later.
if winuser::IsProcessDPIAware() != FALSE {
    // If the process is DPI aware, then scaling must be handled by the application using
    // this DPI value.
    use std::time::Instant;
    use std::thread;
    use winapi::um::wingdi::LOGPIXELSY;
    use winapi::um::errhandlingapi::{SetLastError, GetLastError};
    let now = Instant::now();
    let thread_id = thread::current().name().map(|n| n.to_string());
    SetLastError(0);
    let sx = GetDeviceCaps(hdc, LOGPIXELSX);
    println!("{:?} getDeviceCaps({:?}, LOGPIXELSX)={:?}, thread {:?}", now, hdc, sx, thread_id);
    if sx == 0 {
        println!("{:?} GetLastError()={:?}, thread {:?}", now, GetLastError(), thread_id);
        let sx = GetDeviceCaps(hdc, LOGPIXELSX);
        println!("{:?} repeat getDeviceCaps({:?}, LOGPIXELSX)={:?}, thread {:?}", now, hdc, sx, thread_id);
        panic!("nothing further from {:?}", thread_id);
    }
    sx as u32
} else {
    // If the process is DPI unaware, then scaling is performed by the OS; we thus return
    // 96 (scale factor 1.0) to prevent the window from being re-scaled by both the
    // application and the WM.
    BASE_DPI
}
```

And here is the relevant output for each main and the other thread panicking.

```
Instant { t: 1378293324485 } getDeviceCaps(0x7b01068b, LOGPIXELSX)=96, thread None
Instant { t: 1378293339034 } getDeviceCaps(0x7b01068b, LOGPIXELSX)=96, thread Some("main")
Instant { t: 1378293339571 } getDeviceCaps(0x7b01068b, LOGPIXELSX)=96, thread Some("main")
Instant { t: 1378293339735 } getDeviceCaps(0x7b01068b, LOGPIXELSX)=96, thread Some("main")
Instant { t: 1378293339734 } getDeviceCaps(0x7b01068b, LOGPIXELSX)=0, thread None
Instant { t: 1378293339734 } GetLastError()=0, thread None
Instant { t: 1378293339734 } repeat getDeviceCaps(0x7b01068b, LOGPIXELSX)=96, thread None
thread '<unnamed>' panicked at 'nothing further from None', C:\Users\sinesc\.cargo\registry\src\github.com-1ecc6299db9ec823\winit-0.16.2\src\platform\windows\dpi.rs:190:17
Instant { t: 1378293404216 } getDeviceCaps(0x7b01068b, LOGPIXELSX)=96, thread Some("main")
Instant { t: 1378293405629 } getDeviceCaps(0x7b01068b, LOGPIXELSX)=96, thread Some("main")
```

```
Instant { t: 1381977772142 } getDeviceCaps(0xffffffffa801114d, LOGPIXELSX)=96, thread Some("main")
Instant { t: 1381977820562 } getDeviceCaps(0xffffffffa801114d, LOGPIXELSX)=96, thread None
Instant { t: 1381977835138 } getDeviceCaps(0xffffffffa801114d, LOGPIXELSX)=96, thread Some("main")
Instant { t: 1381977835897 } getDeviceCaps(0xffffffffa801114d, LOGPIXELSX)=96, thread None
Instant { t: 1381977835896 } getDeviceCaps(0xffffffffa801114d, LOGPIXELSX)=0, thread Some("main")
Instant { t: 1381977835896 } GetLastError()=0, thread Some("main")
Instant { t: 1381977835896 } repeat getDeviceCaps(0xffffffffa801114d, LOGPIXELSX)=96, thread Some("main")
thread 'main' panicked at 'nothing further from Some("main")', C:\Users\sinesc\.cargo\registry\src\github.com-1ecc6299db9ec823\winit-0.16.2\src\platform\windows\dpi.rs:190:17
Instant { t: 1381977882343 } getDeviceCaps(0xffffffffa801114d, LOGPIXELSX)=96, thread None
Instant { t: 1381977897496 } getDeviceCaps(0xffffffffa801114d, LOGPIXELSX)=96, thread None
```

**Comment by francesca64** [comment]
Hm, since DPI is global and constant with the `GetDeviceCaps` API, it's probably a good idea for us to just query it once from the internal event loop thread and tuck that into a `lazy_static` or whatever.

I'm not super sure what's happening here, but my default assumption with Windows functions is that they have particular threading requirements.

**Comment by aloucks** [comment]
Could this be the issue?

> Note that the handle to the DC can only be used by a single thread at any one time.

https://docs.microsoft.com/en-us/windows/desktop/api/winuser/nf-winuser-getdc

It looks like the `hdc` is cached for the lifetime of the window:
https://github.com/tomaka/winit/blob/040d3f5d8b0f56e2cbd89e98d7cb4b0b0f83f6d1/src/platform/windows/window.rs#L981-L987

Maybe cache the dpi instead of `hdc`:

https://github.com/tomaka/winit/blob/040d3f5d8b0f56e2cbd89e98d7cb4b0b0f83f6d1/src/platform/windows/window.rs#L1001

**Comment by sinesc** [comment]
> Could this be the issue?

This seemed promising, so instead of using the provided `hdc` I obtained a new one with `let hdc = winuser::GetDC(hwnd);`.

This still panicked, but interestingly because GetDC() returned a null pointer now. Required a lot more violent mouse-shaking to get it to crash though...

**Comment by aloucks** [comment]
I think @francesca64 was saying that the dpi never changes. What happens if you change `WindowWrrapper(HWND, HDC)` to `WindowWrapper(HWND, f64)` or `WindowWrapper { hwnd: HWND, dpi: f64 }` and save the result of `get_window_dpi` instead of the `hdc` itself. I don't see any usage of `hdc` outside of fetching the window dpi and this way you're only calling `GetDC` once inside the window initialization.

**Comment by kryptan** [comment]
Making unnecessary syscalls on every frame is probably not a very good idea. In this particular case glium [does it](https://github.com/glium/glium/blob/37a80c854ee2caed636c0dc493dcf38146e36e42/src/backend/glutin/mod.rs#L260). It calls both `get_hidpi_factor()` and `get_inner_size()` which internally also calls `get_hidpi_factor()`. When I am writing apps using winit I always cache windows size and hidpi factor and change them when winit informs me that they are changing. Maybe winit should do this caching itself so methods like `get_hidpi_factor()` and `get_inner_size()` are cheap and always simply return cached value?

**Comment by francesca64** [comment]
@kryptan that's how `get_hidpi_factor` is on X11, and off the top of my head it would be trivial to do the same on Windows and macOS. There aren't really any technical obstacles to `get_inner_size` working the same way, *but* prior to the fabled `EventLoop` 2.0 I'm not sure how kosher it would be for us to rely on the assumption that the event loop is actually running.

**Comment by sinesc** [comment]
WindowState already has a DPI-factor. Any reason not to use that in get_hidpi_factor()? I have a patch that does that, removes HDC from WindowWrapper and refactors usages. Should I send a PR?

```rust
// This is different from the value in `SavedWindowInfo`! That one represents the DPI saved upon entering
// fullscreen. This will always be the most recent DPI for the window.
pub dpi_factor: f64,
```

`get_hidpi_factor`  would then become:

```rust
#[inline]
pub fn get_hidpi_factor(&self) -> f64 {
    self.window_state.lock().unwrap().dpi_factor
}
```

**Comment by francesca64** [comment]
Yeah, that sounds good!

**Comment by francesca64** [comment]
Fixed via #618
